File: c++/2023/solutions/day03-1.cpp
#include <iostream>
#include <fstream>
#include <vector>

struct PartNumber {
	int i;
	int j;
	int length;
	std::string num;
};

bool isDigit(char c) {
	return c >= '0' && c <= '9';
}

int main() {
	std::ifstream input{ "2023/inputs/day03-input.txt" };
	if (!input.is_open()) std::cout << "file did not open\n";
	
	std::vector<std::string> engine{};
	std::string s;
	while (std::getline(input, s)) engine.push_back(s);
	
	std::vector<PartNumber> numbers{};
	for (int i = 0; i < engine.size(); ++i) {
		for (int j = 0; j < engine[i].size(); ++j) {
			if (isDigit(engine[i][j])) {
				PartNumber p{ i, j, 0 };
				p.num += engine[i][j];
				int length{ 1 };
				while (j + length < engine[i].size() && isDigit(engine[i][j+length])) {
					p.num += engine[i][j+length];
					++length;
				}
				p.length = length;
				numbers.push_back(p);
				
				j = j + length;
			}
		}
	}
	
	int sum = 0;
	for (PartNumber& p : numbers) {
		for (int i = p.i - 1; i <= p.i + 1; ++i) {
			for (int j = p.j - 1; j <= p.j + p.length; ++j) {
				if (i < 0 || j < 0 || i >= engine.size() || j >= engine[i].size()
					|| (i == p.i && j >= p.j && j < p.j + p.length)) continue;
				if (engine[i][j] != '.') {
					sum += std::stoi(p.num);
					break;
				}
			}
		}
	}
	
	std::cout << sum << '\n';
}

File: haskell/2023/solutions/day03-1.hs
import Data.Char as C

type Coord = (Int, Int)
data PartNum = PartNum Coord Int Int deriving (Eq, Show)

inputChar :: [String] -> Coord -> Char
inputChar grid (x, y) = (grid !! y) !! x

partNums = []
getNums :: [String] -> [PartNum]
getNums grid = reverse $ helper grid 0 0 []
   where helper grid x y acc
            | y >= length grid        = acc
            | x >= length (grid !! y) = helper grid 0 (y + 1) acc
            | len > 0        = helper grid (x + len) y ((PartNum (x, y) len (read taken)):acc)
            | otherwise      = helper grid (x + 1) y acc
            where taken = takeWhile isDigit (drop x (grid !! y))
                  len = length taken

surrounds :: [String] -> PartNum -> [Char]
surrounds grid (PartNum (x, y) len _) =
   [inputChar grid (x1, y1) | x1 <- [x-1..x+len], y1 <- [y-1..y+1], 
               y1 >= 0 && y1 < length grid, x1 >= 0 && x1 < length (grid !! y1),
               y1 /= y || x1 < x || x1 >= x + len]

isPartNum :: [String] -> PartNum -> Bool
isPartNum grid p = any (/='.') (surrounds grid p)

solution :: String -> String
solution x = show $ sum $ map (\(PartNum (_,_) _ n) -> n)$ filter (isPartNum grid) (getNums grid)
   where grid = lines x

main = do
       file <- readFile "../inputs/2023/day03-input.txt"
       putStrLn $ solution file